### PR TITLE
Add algorithm reduce sum kernel

### PR DESCRIPTION
**Pull request by MrBurmark**
# Add an algorithm reduce sum kernel.

Add an algorithm reduce sum kernel so we can more easily compare reduction implementations. This has gpu tunings for a native library like cub/rocprim as well as the normal perf suite base implementation and raja reducer implemention.
This also serves as an example of how to implement multiple tunings for kernels.

- This PR is a (refactoring, bugfix, feature, something else)
- It does the following (modify list as needed):
  - Modifies/refactors (class or method) (how?)
  - Fixes (issue number(s))
  - Adds (specific feature) at the request of (project or person)

### CODE REVIEW

File: src/algorithm/REDUCE_SUM-Cuda.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    stopTimer();
+
+    REDUCE_SUM_DATA_TEARDOWN_CUDA;
+
+  } else {
+     getCout() << "\n  REDUCE_SUM : Unknown Cuda variant id = " << vid << std::endl;
+  }
+}
+
+void REDUCE_SUM::runCudaVariant(VariantID vid, size_t tune_idx)
```
Comment (rhornung67) [inline src/algorithm/REDUCE_SUM-Cuda.cpp:201]: Please insert some blank lines between code blocks to make it easier to parse visually

File: src/algorithm/REDUCE_SUM-Cuda.cpp [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        t += 1;
+      }
+    });
+  } else {
+     getCout() << "\n  REDUCE_SUM : Unknown Cuda variant id = " << vid << std::endl;
+  }
+}
+
+void REDUCE_SUM::setCudaTuningDefinitions(VariantID vid)
+{
```
Comment (rhornung67) [inline src/algorithm/REDUCE_SUM-Cuda.cpp:235]: Ditto about blank lines.

File: src/algorithm/REDUCE_SUM.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  void runSeqVariant(VariantID vid, size_t tune_idx);
+  void runOpenMPVariant(VariantID vid, size_t tune_idx);
+  void runCudaVariant(VariantID vid, size_t tune_idx);
+  void runHipVariant(VariantID vid, size_t tune_idx);
+  void runOpenMPTargetVariant(VariantID vid, size_t tune_idx);
+
+  void setCudaTuningDefinitions(VariantID vid);
+  void setHipTuningDefinitions(VariantID vid);
+  void runCudaVariantCub(VariantID vid);
```
Comment (rhornung67) [inline src/algorithm/REDUCE_SUM.hpp:64]: It's not clear to me how these variants are executed. Are they chosen as replacements for the 'standard' variants via command line args?

Comment (rhornung67) [reply src/algorithm/REDUCE_SUM.hpp:64]: So the output will show multiple timings, for example, for "sub-variants"?

Comment (MrBurmark) [reply src/algorithm/REDUCE_SUM.hpp:64]: Yes, the output files include info for each "tuning".

**Comment by rhornung67** [comment]
Grrrfff. Corona CI is plugged up again.

**Review by rhornung67 — DISMISSED** [review]

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**